matrix/test_unittest_matrix.py

```diff
@@ -0,0 +1,177 @@
+
+        with self.assertRaises(ValueError):
+            m1 + m3
+
+        with self.assertRaises(AttributeError):
+            m1 + 10
+
+        with self.assertRaises(AttributeError):
+            m1 + 0.5
+
+    def test_sub(self):
+        """Test subtraction operator."""
+        m1 = Matrix(((1, 2), (3, 4)))
+        m2 = Matrix(((5, 6), (7, 8)))
+        m3 = Matrix.ones(3)
+
+        self.assertEqual(m1 - m2, Matrix(((-4, -4), (-4, -4))))
+
+        with self.assertRaises(ValueError):
+            m1 - m3
+
+        with self.assertRaises(AttributeError):
+            m1 - 10
+
+        with self.assertRaises(AttributeError):
+            m1 - 0.5
+
+    def test_mul(self):
+        """Test multiplication and reverse mul operators."""
+        m1 = Matrix(((1, 2), (3, 4)))
+        m2 = Matrix(((5, 6), (7, 8)))
+        m3 = Matrix.ones(4)
+
+        self.assertEqual(m1 * m2, Matrix(((5, 12), (21, 32))))
+        self.assertEqual(m1 * 10, Matrix(((10, 20), (30, 40))))
+        self.assertEqual(m1 * 0.5, Matrix(((0.5, 1), (1.5, 2))))
+
+        # reverse mul
+        self.assertEqual(10 * m1, Matrix(((10, 20), (30, 40))))
+        self.assertEqual(0.5 * m1, Matrix(((0.5, 1), (1.5, 2))))
+
+        with self.assertRaises(ValueError):
+            m1 * m3
+
+    def test_div(self):
+        """Test true division operator operator."""
+        m1 = Matrix(((1, 2), (3, 4)))
+        m2 = Matrix(((5, 6), (7, 8)))
+        m3 = Matrix.ones(4)
+
+        self.assertEqual(m1 / 2, Matrix(((0.5, 1), (1.5, 2))))
+        self.assertEqual(m1 / 0.5, Matrix(((2, 4), (6, 8))))
+
+        with self.assertRaises(TypeError):
+            m1 / m2
+
+        with self.assertRaises(TypeError):
+            m1 / m3
+
+        with self.assertRaises(TypeError):
+            m1 / (1, 2)
+
+    def test_equal(self):
+        """Test equal boolean operator."""
+        self.assertEqual(
+            Matrix.unity(3), Matrix(((1, 0, 0), (0, 1, 0), (0, 0, 1))))
+
+        self.assertEqual(
+            Matrix.ones(2), Matrix(((1, 1), (1, 1))))
+
+    def test_not_equal(self):
+        """Test not equal boolean operator."""
+        self.assertNotEqual(Matrix(((1, 2), (3, 4))), Matrix(((5, 6), (7, 8))))
+        self.assertNotEqual(Matrix(((1, 2), (3, 4))), Matrix.ones(3))
+
+
+class TestMatrixIteration(unittest.TestCase):
+    """Test iteration support of Matrix."""
+    def test_iteration(self):
+        """Test __iter__ function."""
+        self.assertEqual(tuple(Matrix.ones(2)), ((1, 1), (1, 1)))
+
+
+class TestMatrixHashable(unittest.TestCase):
+    """Test hash support of Matrix."""
+    def test_hash(self):
+        """Test __hash__ function and uses dictionary keys."""
+        d = {Matrix(((1, 1), (2, 2))): 2, Matrix(((1, 1), (2, 3))): 3}
+
+        self.assertEqual(d, {Matrix(((1, 1), (2, 2))): 2,
+                             Matrix(((1, 1), (2, 3))): 3})
+
+        self.assertNotEqual(hash(Matrix.ones(2)), hash(Matrix.unity(2)))
+
+
```

Comment:
what about testsuite?

Reply:
@rfire01 I'm not sure testsuites are really interesting
No one creates them anymore, all test-engines can discover all tests anyway

Reply:
So should I implement it?

Reply:
just understand what is it, no need to implement